File: android/ToneRewriterImplementation.kt
package com.lorapok.keyboard

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.util.TypedValue
import android.view.Gravity
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup

/**
 * Lorapok Keuboard — Tone Rewriter Implementation (Android)
 *
 * Provides 8 Bengali tone transformations:
 *   Free:    Formal, Friendly, Respectful, Concise, Detailed
 *   Premium: Poetic, Humorous, Empathetic
 *
 * Uses a hybrid approach:
 *   - Rule-based engine for pronoun/verb/formality transformations (<100ms)
 *   - Neural engine placeholder for complex tones (<500ms)
 */

// ══════════════════════════════════════════════════
// Data Models
// ══════════════════════════════════════════════════

enum class ToneType(
    val id: String,
    val bengaliName: String,
    val emoji: String,
    val isPremium: Boolean
) {
    FORMAL("formal", "আনুষ্ঠানিক", "🎩", false),
    FRIENDLY("friendly", "বন্ধুত্বপূর্ণ", "😊", false),
    RESPECTFUL("respectful", "শ্রদ্ধাশীল", "🙏", false),
    CONCISE("concise", "সংক্ষিপ্ত", "⚡", false),
    DETAILED("detailed", "বিস্তারিত", "📝", false),
    POETIC("poetic", "কাব্যিক", "📚", true),
    HUMOROUS("humorous", "হাস্যরসাত্মক", "😄", true),
    EMPATHETIC("empathetic", "সহানুভূতিশীল", "💝", true);

    val displayName get() = "$emoji $bengaliName"
}

data class ToneResult(
    val originalText: String,
    val rewrittenText: String,
    val tone: ToneType,
    val confidence: Float,
    val processingTimeMs: Long
)

// ══════════════════════════════════════════════════
// Bengali Tone Rules (Rule-Based Engine)
// ══════════════════════════════════════════════════

object BengaliToneRules {

    // Pronoun conversion tables
    private val pronounUpgrade = mapOf(
        "তুই" to "তুমি", "তুমি" to "আপনি", "তোর" to "তোমার",
        "তোমার" to "আপনার", "তোকে" to "তোমাকে", "তোমাকে" to "আপনাকে",
        "তোরা" to "তোমরা", "তোমরা" to "আপনারা"
    )

    private val pronounDowngrade = mapOf(
        "আপনি" to "তুমি", "আপনার" to "তোমার", "আপনাকে" to "তোমাকে",
        "আপনারা" to "তোমরা", "তুমি" to "তুই", "তোমার" to "তোর",
        "তোমাকে" to "তোকে", "তোমরা" to "তোরা"
    )

    // Verb formality mapping
    private val verbFormal = mapOf(
        "আসবে" to "উপস্থিত থাকবেন", "যাবে" to "যাবেন",
        "করবে" to "করবেন", "বলবে" to "বলবেন",
        "দেবে" to "দেবেন", "খাবে" to "খাবেন",
        "পারবে" to "পারবেন", "আসো" to "আসুন",
        "যাও" to "যান", "করো" to "করুন",
        "বলো" to "বলুন", "দাও" to "দিন",
        "খাও" to "খান", "বসো" to "বসুন",
        "আসবি" to "আসবেন", "যাবি" to "যাবেন",
        "করবি" to "করবেন", "পারবি" to "পারবেন"
    )

    private val verbCasual = mapOf(
        "উপস্থিত থাকবেন" to "আসবে", "যাবেন" to "যাবে",
        "করবেন" to "করবে", "বলবেন" to "বলবে",
        "দেবেন" to "দেবে", "খাবেন" to "খাবে",
        "পারবেন" to "পারবে", "আসুন" to "আসো",
        "যান" to "যাও", "করুন" to "করো",
        "বলুন" to "বলো", "দিন" to "দাও"
    )

    // Word formality mapping
    private val wordFormal = mapOf(
        "কাল" to "আগামীকাল", "আজকে" to "আজ",
        "এখনি" to "এই মুহূর্তে", "অফিস" to "কার্যালয়",
        "মিটিং" to "সভা", "বস" to "ঊর্ধ্বতন কর্মকর্তা",
        "ওকে" to "ঠিক আছে", "হ্যাঁ" to "জি",
        "না" to "না, ধন্যবাদ"
    )

    private val wordCasual = mapOf(
        "আগামীকাল" to "কাল", "এই মুহূর্তে" to "এখনি",
        "কার্যালয়" to "অফিস", "সভা" to "মিটিং",
        "ঊর্ধ্বতন কর্মকর্তা" to "বস", "জি" to "হ্যাঁ"
    )

    // Filler words to remove for concise tone
    private val fillerWords = setOf(
        "আমি মনে করি", "সম্ভবত", "হয়তো", "আসলে",
        "যদি বলি", "মূলত", "বস্তুত", "প্রকৃতপক্ষে",
        "অর্থাৎ", "যাই হোক"
    )

    // Respectful prefixes/suffixes
    private val respectfulAdditions = mapOf(
        "আসতে" to "পধারতে", "যেতে" to "যেতে",
        "বলতে" to "বলতে", "দেখতে" to "দেখতে"
    )

    fun applyPronounUpgrade(text: String): String {
        var result = text
        for ((from, to) in pronounUpgrade) {
            result = result.replace(from, to)
        }
        return result
    }

    fun applyPronounDowngrade(text: String): String {
        var result = text
        for ((from, to) in pronounDowngrade) {
            result = result.replace(from, to)
        }
        return result
    }

    fun applyVerbFormalize(text: String): String {
        var result = text
        for ((from, to) in verbFormal) {
            result = result.replace(from, to)
        }
        return result
    }

    fun applyVerbCasualize(text: String): String {
        var result = text
        for ((from, to) in verbCasual) {
            result = result.replace(from, to)
        }
        return result
    }

    fun applyWordFormalize(text: String): String {
        var result = text
        for ((from, to) in wordFormal) {
            result = result.replace(from, to)
        }
        return result
    }

    fun applyWordCasualize(text: String): String {
        var result = text
        for ((from, to) in wordCasual) {
            result = result.replace(from, to)
        }
        return result
    }

    fun removeFillers(text: String): String {
        var result = text
        for (filler in fillerWords) {
            result = result.replace(filler, "")
        }
        return result.replace(Regex("\\s+"), " ").trim()
    }
}

// ══════════════════════════════════════════════════
// Tone Rewriting Engine
// ══════════════════════════════════════════════════

class ToneRewritingEngine(private val context: Context) {

    fun rewrite(text: String, targetTone: ToneType): ToneResult {
        val startTime = System.currentTimeMillis()

        val rewritten = when (targetTone) {
            ToneType.FORMAL -> toFormal(text)
            ToneType.FRIENDLY -> toFriendly(text)
            ToneType.RESPECTFUL -> toRespectful(text)
            ToneType.CONCISE -> toConcise(text)
            ToneType.DETAILED -> toDetailed(text)
            ToneType.POETIC -> toPoetic(text)
            ToneType.HUMOROUS -> toHumorous(text)
            ToneType.EMPATHETIC -> toEmpathetic(text)
        }

        val elapsed = System.currentTimeMillis() - startTime

        return ToneResult(
            originalText = text,
            rewrittenText = rewritten,
            tone = targetTone,
            confidence = 0.85f,
            processingTimeMs = elapsed
        )
    }

    private fun toFormal(text: String): String {
        var result = BengaliToneRules.applyPronounUpgrade(text)
        result = BengaliToneRules.applyVerbFormalize(result)
        result = BengaliToneRules.applyWordFormalize(result)
        // Add polite question marker if question
        if (result.contains("?") || result.contains("কি")) {
            result = result.replace("কি", "কি দয়া করে")
        }
        return result
    }

    private fun toFriendly(text: String): String {
        var result = BengaliToneRules.applyPronounDowngrade(text)
        result = BengaliToneRules.applyVerbCasualize(result)
        result = BengaliToneRules.applyWordCasualize(result)
        // Add friendly emoji
        if (!result.contains("😊")) result += " 😊"
        return result
    }

    private fun toRespectful(text: String): String {
        var result = BengaliToneRules.applyPronounUpgrade(text)
        result = BengaliToneRules.applyVerbFormalize(result)
        // Apply respectful verb forms
        result = result.replace("আসতে", "পধারতে")
        // Add respectful prefix
        if (!result.startsWith("দয়া করে") && !result.startsWith("অনুগ্রহ করে")) {
            result = "অনুগ্রহ করে $result"
        }
        return result
    }

    private fun toConcise(text: String): String {
        var result = BengaliToneRules.removeFillers(text)
        // Shorten common phrases
        result = result.replace("উপস্থিত থাকবেন", "আসবেন")
        result = result.replace("এই মুহূর্তে", "এখন")
        result = result.replace(Regex("\\s+"), " ").trim()
        return result
    }

    private fun toDetailed(text: String): String {
        // Add elaboration markers
        var result = text
        if (!result.contains("অর্থাৎ") && result.length < 50) {
            result += "। বিস্তারিত বলতে গেলে, এটি একটি গুরুত্বপূর্ণ বিষয়"
        }
        return result
    }

    private fun toPoetic(text: String): String {
        // Simple poetic word substitutions (neural engine would do better)
        var result = text
        result = result.replace("সুন্দর", "রূপের মাধুরী")
        result = result.replace("আবহাওয়া", "প্রকৃতি")
        result = result.replace("ভালো", "মনোরম")
        result = result.replace("দেখা", "দর্শন")
        result = result.replace("খুব", "অতীব")
        return result
    }

    private fun toHumorous(text: String): String {
        var result = text
        // Add playful markers
        if (!result.endsWith("😄") && !result.endsWith("😂")) {
            result += " 😄"
        }
        return result
    }

    private fun toEmpathetic(text: String): String {
        var result = text
        // Add empathetic prefix
        if (!result.startsWith("আমি বুঝতে পারছি")) {
            result = "আমি বুঝতে পারছি। $result"
        }
        if (!result.endsWith("💝")) result += " 💝"
        return result
    }
}

// ══════════════════════════════════════════════════
// Tone Rewriter View (UI Component)
// ══════════════════════════════════════════════════

class ToneRewriterView(private val context: Context) {

    private val engine = ToneRewritingEngine(context)
    private var currentResult: ToneResult? = null
    private var onAcceptListener: ((String) -> Unit)? = null

    fun setOnAcceptListener(listener: (String) -> Unit) {
        onAcceptListener = listener
    }

    /**
     * Show the tone rewriter bottom sheet dialog.
     */
    fun show(selectedText: String) {
        val dialog = BottomSheetDialog(context)
        val rootLayout = buildDialogUI(selectedText, dialog)
        dialog.setContentView(rootLayout)
        dialog.show()
    }

    private fun buildDialogUI(text: String, dialog: BottomSheetDialog): LinearLayout {
        val root = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(dp(16), dp(16), dp(16), dp(16))
            setBackgroundColor(Color.WHITE)
        }

        // Title
        root.addView(TextView(context).apply {
            this.text = "✨ টোন পরিবর্তন করুন"
            setTextSize(TypedValue.COMPLEX_UNIT_SP, 20f)
            setTextColor(Color.parseColor("#1F1F1F"))
            setPadding(0, 0, 0, dp(8))
        })

        // Original text
        root.addView(TextView(context).apply {
            this.text = "মূল: $text"
            setTextSize(TypedValue.COMPLEX_UNIT_SP, 14f)
            setTextColor(Color.parseColor("#666666"))
            setPadding(0, 0, 0, dp(12))
        })

        // Tone chips
        val chipGroup = ChipGroup(context).apply {
            isSingleSelection = true
            setPadding(0, 0, 0, dp(12))
        }

        for (tone in ToneType.values()) {
            val chip = Chip(context).apply {
                this.text = tone.displayName
                isCheckable = true
                isEnabled = !tone.isPremium // Disable premium tones (for now)
                setOnCheckedChangeListener { _, isChecked ->
                    if (isChecked) {
                        val result = engine.rewrite(text, tone)
                        currentResult = result
                        updatePreview(root, result)
                    }
                }
            }
            chipGroup.addView(chip)
        }
        root.addView(chipGroup)

        // Preview area (placeholder)
        root.addView(TextView(context).apply {
            tag = "preview_text"
            this.text = "☝️ উপরে একটি টোন নির্বাচন করুন"
            setTextSize(TypedValue.COMPLEX_UNIT_SP, 16f)
            setTextColor(Color.parseColor("#1A73E8"))
            setPadding(dp(12), dp(12), dp(12), dp(12))
            background = GradientDrawable().apply {
                setColor(Color.parseColor("#F0F4FF"))
                cornerRadius = dp(8).toFloat()
            }
        })

        // Buttons row
        val buttonRow = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.END
            setPadding(0, dp(16), 0, 0)
        }

        buttonRow.addView(Button(context).apply {
            this.text = "বাতিল"
            setOnClickListener { dialog.dismiss() }
        })

        buttonRow.addView(Button(context).apply {
            this.text = "ব্যবহার করুন"
            setTextColor(Color.WHITE)
            background = GradientDrawable().apply {
                setColor(Color.parseColor("#1A73E8"))
                cornerRadius = dp(8).toFloat()
            }
            setOnClickListener {
                currentResult?.let { result ->
                    onAcceptListener?.invoke(result.rewrittenText)
                }
                dialog.dismiss()
            }
        })

        root.addView(buttonRow)
        return root
    }

    private fun updatePreview(root: LinearLayout, result: ToneResult) {
        val preview = root.findViewWithTag<TextView>("preview_text")
        preview?.text = "${result.tone.emoji} ${result.rewrittenText}\n\n⏱️ ${result.processingTimeMs}ms"
    }

    private fun dp(value: Int): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, value.toFloat(),
            context.resources.displayMetrics
        ).toInt()
    }
}

File: ios/ToneRewriterImplementation.swift
import UIKit
import SwiftUI

// MARK: - Lorapok Keuboard — iOS Tone Rewriter Implementation
// 8 Bengali tone transformations with rule-based engine
// Min iOS: 13.0 | Memory: <48 MB

// MARK: - Tone Type

enum ToneType: String, CaseIterable {
    case formal = "formal"
    case friendly = "friendly"
    case respectful = "respectful"
    case concise = "concise"
    case detailed = "detailed"
    case poetic = "poetic"
    case humorous = "humorous"
    case empathetic = "empathetic"

    var bengaliName: String {
        switch self {
        case .formal: return "আনুষ্ঠানিক"
        case .friendly: return "বন্ধুত্বপূর্ণ"
        case .respectful: return "শ্রদ্ধাশীল"
        case .concise: return "সংক্ষিপ্ত"
        case .detailed: return "বিস্তারিত"
        case .poetic: return "কাব্যিক"
        case .humorous: return "হাস্যরসাত্মক"
        case .empathetic: return "সহানুভূতিশীল"
        }
    }

    var emoji: String {
        switch self {
        case .formal: return "🎩"
        case .friendly: return "😊"
        case .respectful: return "🙏"
        case .concise: return "⚡"
        case .detailed: return "📝"
        case .poetic: return "📚"
        case .humorous: return "😄"
        case .empathetic: return "💝"
        }
    }

    var isPremium: Bool {
        return [.poetic, .humorous, .empathetic].contains(self)
    }

    var displayName: String { "\(emoji) \(bengaliName)" }
}

// MARK: - Tone Result

struct ToneResult {
    let originalText: String
    let rewrittenText: String
    let tone: ToneType
    let confidence: Float
    let processingTimeMs: Int64
}

// MARK: - Bengali Tone Rules

struct BengaliToneRules {

    static let pronounUpgrade: [String: String] = [
        "তুই": "তুমি", "তুমি": "আপনি", "তোর": "তোমার",
        "তোমার": "আপনার", "তোকে": "তোমাকে", "তোমাকে": "আপনাকে",
        "তোরা": "তোমরা", "তোমরা": "আপনারা"
    ]

    static let pronounDowngrade: [String: String] = [
        "আপনি": "তুমি", "আপনার": "তোমার", "আপনাকে": "তোমাকে",
        "আপনারা": "তোমরা", "তুমি": "তুই", "তোমার": "তোর"
    ]

    static let verbFormal: [String: String] = [
        "আসবে": "উপস্থিত থাকবেন", "যাবে": "যাবেন",
        "করবে": "করবেন", "বলবে": "বলবেন",
        "পারবে": "পারবেন", "আসো": "আসুন",
        "যাও": "যান", "করো": "করুন",
        "আসবি": "আসবেন", "পারবি": "পারবেন"
    ]

    static let verbCasual: [String: String] = [
        "উপস্থিত থাকবেন": "আসবে", "যাবেন": "যাবে",
        "করবেন": "করবে", "বলবেন": "বলবে",
        "পারবেন": "পারবে", "আসুন": "আসো",
        "যান": "যাও", "করুন": "করো"
    ]

    static let wordFormal: [String: String] = [
        "কাল": "আগামীকাল", "অফিস": "কার্যালয়",
        "মিটিং": "সভা", "হ্যাঁ": "জি"
    ]

    static let wordCasual: [String: String] = [
        "আগামীকাল": "কাল", "কার্যালয়": "অফিস",
        "সভা": "মিটিং", "জি": "হ্যাঁ"
    ]

    static let fillerWords: [String] = [
        "আমি মনে করি", "সম্ভবত", "হয়তো", "আসলে", "মূলত"
    ]

    static func applyReplacements(_ text: String, rules: [String: String]) -> String {
        var result = text
        for (from, to) in rules {
            result = result.replacingOccurrences(of: from, with: to)
        }
        return result
    }

    static func removeFillers(_ text: String) -> String {
        var result = text
        for filler in fillerWords {
            result = result.replacingOccurrences(of: filler, with: "")
        }
        // Collapse whitespace
        result = result.components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }.joined(separator: " ")
        return result
    }
}

// MARK: - Tone Rewriting Engine

class ToneRewritingEngine {

    func rewrite(text: String, targetTone: ToneType) -> ToneResult {
        let start = DispatchTime.now()

        let rewritten: String
        switch targetTone {
        case .formal:
            rewritten = toFormal(text)
        case .friendly:
            rewritten = toFriendly(text)
        case .respectful:
            rewritten = toRespectful(text)
        case .concise:
            rewritten = toConcise(text)
        case .detailed:
            rewritten = toDetailed(text)
        case .poetic:
            rewritten = toPoetic(text)
        case .humorous:
            rewritten = toHumorous(text)
        case .empathetic:
            rewritten = toEmpathetic(text)
        }

        let end = DispatchTime.now()
        let elapsed = Int64(end.uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000

        return ToneResult(
            originalText: text, rewrittenText: rewritten,
            tone: targetTone, confidence: 0.85, processingTimeMs: elapsed
        )
    }

    private func toFormal(_ text: String) -> String {
        var r = BengaliToneRules.applyReplacements(text, rules: BengaliToneRules.pronounUpgrade)
        r = BengaliToneRules.applyReplacements(r, rules: BengaliToneRules.verbFormal)
        r = BengaliToneRules.applyReplacements(r, rules: BengaliToneRules.wordFormal)
        return r
    }

    private func toFriendly(_ text: String) -> String {
        var r = BengaliToneRules.applyReplacements(text, rules: BengaliToneRules.pronounDowngrade)
        r = BengaliToneRules.applyReplacements(r, rules: BengaliToneRules.verbCasual)
        r = BengaliToneRules.applyReplacements(r, rules: BengaliToneRules.wordCasual)
        return r.hasSuffix("😊") ? r : r + " 😊"
    }

    private func toRespectful(_ text: String) -> String {
        var r = BengaliToneRules.applyReplacements(text, rules: BengaliToneRules.pronounUpgrade)
        r = BengaliToneRules.applyReplacements(r, rules: BengaliToneRules.verbFormal)
        r = r.replacingOccurrences(of: "আসতে", with: "পধারতে")
        if !r.hasPrefix("অনুগ্রহ করে") { r = "অনুগ্রহ করে " + r }
        return r
    }

    private func toConcise(_ text: String) -> String {
        var r = BengaliToneRules.removeFillers(text)
        r = r.replacingOccurrences(of: "উপস্থিত থাকবেন", with: "আসবেন")
        return r
    }

    private func toDetailed(_ text: String) -> String {
        if text.count < 50 {
            return text + "। বিস্তারিত বলতে গেলে, এটি একটি গুরুত্বপূর্ণ বিষয়"
        }
        return text
    }

    private func toPoetic(_ text: String) -> String {
        var r = text
        r = r.replacingOccurrences(of: "সুন্দর", with: "রূপের মাধুরী")
        r = r.replacingOccurrences(of: "আবহাওয়া", with: "প্রকৃতি")
        r = r.replacingOccurrences(of: "ভালো", with: "মনোরম")
        r = r.replacingOccurrences(of: "খুব", with: "অতীব")
        return r
    }

    private func toHumorous(_ text: String) -> String {
        return text.hasSuffix("😄") ? text : text + " 😄"
    }

    private func toEmpathetic(_ text: String) -> String {
        var r = text
        if !r.hasPrefix("আমি বুঝতে পারছি") { r = "আমি বুঝতে পারছি। " + r }
        return r.hasSuffix("💝") ? r : r + " 💝"
    }
}

// MARK: - SwiftUI Tone Rewriter View

@available(iOS 14.0, *)
struct ToneRewriterView: View {
    let originalText: String
    let onAccept: (String) -> Void
    let onDismiss: () -> Void

    @State private var selectedTone: ToneType? = nil
    @State private var result: ToneResult? = nil

    private let engine = ToneRewritingEngine()

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            // Title
            Text("✨ টোন পরিবর্তন করুন")
                .font(.title2).bold()

            // Original
            Text("মূল: \(originalText)")
                .font(.subheadline).foregroundColor(.secondary)

            // Tone chips
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(ToneType.allCases, id: \.rawValue) { tone in
                        Button(action: {
                            selectedTone = tone
                            result = engine.rewrite(text: originalText, targetTone: tone)
                        }) {
                            Text(tone.displayName)
                                .font(.caption)
                                .padding(.horizontal, 12).padding(.vertical, 6)
                                .background(selectedTone == tone ? Color.blue.opacity(0.2) : Color.gray.opacity(0.1))
                                .cornerRadius(16)
                        }
                        .disabled(tone.isPremium)
                        .opacity(tone.isPremium ? 0.5 : 1.0)
                    }
                }
            }

            // Preview
            if let r = result {
                VStack(alignment: .leading, spacing: 4) {
                    Text("\(r.tone.emoji) পরিবর্তিত:")
                        .font(.caption).foregroundColor(.blue)
                    Text(r.rewrittenText)
                        .font(.body).bold()
                    Text("⏱️ \(r.processingTimeMs)ms")
                        .font(.caption2).foregroundColor(.secondary)
                }
                .padding(12)
                .background(Color.blue.opacity(0.05))
                .cornerRadius(8)
            } else {
                Text("☝️ উপরে একটি টোন নির্বাচন করুন")
                    .foregroundColor(.secondary).padding(12)
            }

            // Buttons
            HStack {
                Spacer()
                Button("বাতিল") { onDismiss() }
                    .foregroundColor(.secondary)
                Button("ব্যবহার করুন") {
                    if let r = result { onAccept(r.rewrittenText) }
                }
                .padding(.horizontal, 16).padding(.vertical, 8)
                .background(Color.blue).foregroundColor(.white)
                .cornerRadius(8)
            }
        }
        .padding(16)
    }
}

// MARK: - UIKit ViewController Wrapper

class ToneRewriterViewController: UIViewController {

    var originalText: String = ""
    var onAccept: ((String) -> Void)?

    private let engine = ToneRewritingEngine()
    private var currentResult: ToneResult?

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        setupUI()
    }

    private func setupUI() {
        let titleLabel = UILabel()
        titleLabel.text = "✨ টোন পরিবর্তন করুন"
        titleLabel.font = .boldSystemFont(ofSize: 20)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(titleLabel)

        NSLayoutConstraint.activate([
            titleLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
            titleLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16)
        ])
    }
}
